src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1793 @@
+                    && dominates(&domtree, liveout_vreg_def_block, *safepoint_block)
+                {
+                    if self.func.block_params(liveout_vreg_def_block).contains(liveout_vreg) {
+                        // Since block params aren't explicitly defined, they are marked as defined
+                        // in the first instruction in the block, even though they are actually
+                        // defined just before that.
+                        // This is the reason why <= is used here instead of just <.
+                        if liveout_vreg_def_inst <= *safepoint_inst {
+                            trace!("Liveout vreg inst: {:?}", self.liveout_vreg_def_inst[liveout_vreg.vreg()]);
+                            trace!("Safepoint inst: {:?}", safepoint_inst);
+                            trace!("Adding a stackmap slot for liveout vreg {:?}", liveout_vreg);
+                            self.safepoint_slots.push((
+                                ProgPoint::before(*safepoint_inst),
+                                Allocation::stack(self.vreg_spillslots[liveout_vreg.vreg()])
+                            ));
+                        }
+                    }
+                    // The definition of the vreg must come before the safepoint instruction
+                    // This is necessary because, while the `dominates` call checks for different
+                    // blocks, in the case where the vreg definition and the safepoint instructions 
+                    // are in the same block, we need to make this check.
+                    else if liveout_vreg_def_inst < *safepoint_inst {
+                        self.safepoint_slots.push((
+                            ProgPoint::before(*safepoint_inst),
+                            Allocation::stack(self.vreg_spillslots[liveout_vreg.vreg()])
+                        ));
+                    }
+                }
+            }
+        }
+        self.safepoint_slots.sort_by(
+            |slot0, slot1| slot0.0.cmp(&slot1.0)
+        );
+    }
+
+    fn alloc_block(&mut self, block: Block) {
+        trace!("{:?} start", block);
+        for inst in self.func.block_insns(block).iter().rev() {
+            self.alloc_inst(block, inst);
+        }
+        self.reload_at_begin(block);
+        trace!("{:?} end\n", block);
+    }
+
+    fn run(&mut self) -> Result<(), RegAllocError> {
+        debug_assert_eq!(self.func.entry_block().index(), 0);
+        for block in (0..self.func.num_blocks()).rev() {
+            self.alloc_block(Block::new(block));
+        }
+        self.build_safepoint_stackmap();
+
+        /////////////////////////////////////////////////////////////////////////////////////
+        trace!("Done!");
+        struct Z(usize);
+        impl std::fmt::Debug for Z {
+            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+                write!(f, "v{}", self.0)
```

Comment:
I think the `VReg` should have an equivalent `Display` implementation, so instead we could push `VReg::new(i, RegClass::Int)` below (class isn't printed by `Display`)